### PR TITLE
:bug:  Snowflake destination: snowflake s3 destination COPY is writing records from different table  in the same raw table fix

### LINKED ISSUE #5664: AWS S3 Staging COPY is writing records from different table in the same raw table
## Enviroment
- **Airbyte version**:  0.29.11-alpha
- **OS Version / Instance**: AWS EC2
- **Deployment**: Docker
- **Source Connector and version**: MySQL 0.4.3
- **Destination Connector and version**: Snowflake 0.3.12
- **Severity**: Critical
- **Step where error happened**: Sync

## Current Behavior
in the `_AIRBYTE_RAW_MYTABLE` i am seeing json blobs in `_AIRBYTE_DATA` column that are pertaining to two different tables.

i can give two examples where this happens. both of which have similar table names
e.g.
* `user` and `user_invites` tables
* `entrydispute` and `entrydisputeevent` tables

in the first table, there are JSON blobs with schemas pertaining to the second table. and so the resulting `_scd` and final tables contain NULL records because `_AIRBYTE_RAW_USER` has a record (row) with raw json data having the schema of `_AIRBYTE_RAW_USER_INVITES` and so the columns in this raw (in the final `_scd` and final table) are NULL because that row doesn't have any data with columns relating to the original table.

It appears that the AWS S3 Staging COPY is copying data from one table into the other

## Expected Behavior
Each table should have `_AIRBYTE_RAW_<TABLE>` with _only_ that table's data

## Logs
*If applicable, please upload the logs from the failing operation. 
For sync jobs, you can download the full logs from the UI by going to the sync attempt page and 
clicking the download logs button at the top right of the logs display window.*

<details>
<summary>LOG</summary> 

```

replace this with
your long log
output here

```

</details>

## Steps to Reproduce
1. I am not entirely sure. I have modified connectors to use standard and then S3 staging (changing this feature back and forth) as well as CDC vs standard (changing this feature back and forth)
2. I have re-used the same source for different connections (and different tables)
3. I have made two of the same connections (one using standard and the other using S3 staging)

## Are you willing to submit a PR?

Remove this with your answer.

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Fix snowflake destination aws s3 Staging  COPY is writing records from different table in the same raw table 

## How
This bug occurs due to the fact that the snowflake does not look for the full path to the file on s3, but uses this path to find all matches in the file path. For example, we have tables `user` and `user_something`. For the table user_something, one file will be found and copying will be done only from it. And for the user table, two files will be found: 1 - _identifier/schema/user_ and 2 - _identifier/schema/user_something_ - because 1 path is part of the second.

I fixed this by prefixing the table name in the file path.

Here are the logs for my test tables (`actor`, `actor_my`, `actor_test` and `actor_payment`) in the case when it works **without a prefix**:

```
2021-09-09 08:40:48 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:48 INFO i.a.i.d.j.c.s.S3StreamCopier(createTemporaryTable):183 - {} - Preparing tmp table in destination for stream: actor, schema: AA_AIRBYTE_TEST_SCHEMA, tmp table name: _airbyte_tmp_suk_actor.
2021-09-09 08:40:49 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:49 INFO i.a.i.d.j.c.s.S3StreamCopier(copyStagingFileToTemporaryTable):189 - {} - Starting copy to tmp table: _airbyte_tmp_suk_actor in destination for stream: actor, schema: AA_AIRBYTE_TEST_SCHEMA, .
2021-09-09 08:40:51 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:51 ERROR i.a.i.d.s.SnowflakeS3StreamCopier(copyS3CsvFileIntoTable):73 - {} - **there was a download from the following files ==>> [{"file":"s3://airbyte-snowflake-integration-tests/692b491d-d39a-4e5b-a120-c44407925017/AA_AIRBYTE_TEST_SCHEMA/actor_test","status":"LOADED","rows_parsed":3,"rows_loaded":3,"error_limit":1,"errors_seen":0}, {"file":"s3://airbyte-snowflake-integration-tests/692b491d-d39a-4e5b-a120-c44407925017/AA_AIRBYTE_TEST_SCHEMA/actor_my","status":"LOADED","rows_parsed":2,"rows_loaded":2,"error_limit":1,"errors_seen":0}, {"file":"s3://airbyte-snowflake-integration-tests/692b491d-d39a-4e5b-a120-c44407925017/AA_AIRBYTE_TEST_SCHEMA/actor","status":"LOADED","rows_parsed":200,"rows_loaded":200,"error_limit":1,"errors_seen":0}, {"file":"s3://airbyte-snowflake-integration-tests/692b491d-d39a-4e5b-a120-c44407925017/AA_AIRBYTE_TEST_SCHEMA/actor_payment","status":"LOADED","rows_parsed":14596,"rows_loaded":14596,"error_limit":1,"errors_seen":0}]**
2021-09-09 08:40:51 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:51 INFO i.a.i.d.j.c.s.S3StreamCopier(copyStagingFileToTemporaryTable):191 - {} - Copy to tmp table _airbyte_tmp_suk_actor in destination for stream actor complete.
2021-09-09 08:40:51 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:51 INFO i.a.i.d.j.c.s.S3StreamCopier(createDestinationTable):197 - {} - Preparing table _airbyte_raw_actor in destination.
2021-09-09 08:40:53 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:40:53 INFO i.a.i.d.j.c.s.S3StreamCopier(createDestinationTable):199 - {} - Table _airbyte_tmp_suk_actor in destination prepared.
```

And my logs for my test tables (`actor`) in the case when it works **with a prefix**:

```
2021-09-09 08:45:55 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:55 INFO i.a.i.d.j.c.s.S3StreamCopier(createDestinationSchema):177 - {} - Creating schema in destination if it doesn't exist: AA_AIRBYTE_TEST_SCHEMA
2021-09-09 08:45:56 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:56 INFO i.a.i.d.j.c.s.S3StreamCopier(createTemporaryTable):183 - {} - Preparing tmp table in destination for stream: actor, schema: AA_AIRBYTE_TEST_SCHEMA, tmp table name: _airbyte_tmp_sot_actor.
2021-09-09 08:45:57 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:57 INFO i.a.i.d.j.c.s.S3StreamCopier(copyStagingFileToTemporaryTable):189 - {} - Starting copy to tmp table: _airbyte_tmp_sot_actor in destination for stream: actor, schema: AA_AIRBYTE_TEST_SCHEMA, .
2021-09-09 08:45:58 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:58 ERROR i.a.i.d.s.SnowflakeS3StreamCopier(copyS3CsvFileIntoTable):73 - {} - **there was a download from the following files ==>> [{"file":"s3://airbyte-snowflake-integration-tests/2db32d83-b50b-4892-97cc-466956a96c34/AA_AIRBYTE_TEST_SCHEMA/bqzactor","status":"LOADED","rows_parsed":200,"rows_loaded":200,"error_limit":1,"errors_seen":0}]**
2021-09-09 08:45:58 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:58 INFO i.a.i.d.j.c.s.S3StreamCopier(copyStagingFileToTemporaryTable):191 - {} - Copy to tmp table _airbyte_tmp_sot_actor in destination for stream actor complete.
2021-09-09 08:45:58 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:45:58 INFO i.a.i.d.j.c.s.S3StreamCopier(createDestinationTable):197 - {} - Preparing table _airbyte_raw_actor in destination.
2021-09-09 08:46:00 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-09-09 08:46:00 INFO i.a.i.d.j.c.s.S3StreamCopier(createDestinationTable):199 - {} - Table _airbyte_tmp_sot_actor in destination prepared.
```

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by etsybaev — APPROVED** [review]
LGTM

**Comment by danieldiamond** [comment]
@andriikorotkov are you able to add to the description? e.g. logs that you've produced or _how_ the SQL for `COPY INTO..` is generated incorrectly
Or is this simply a bug with snowflake when file extension is omitted - and if so, are we considering the future if we want to `COPY INTO` with other extensions/compressions?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3StreamCopier.java [hunk 1]
```diff
@@ -118,6 +155,10 @@ public S3StreamCopier(String stagingFolder,
     }
   }
 
+  private String prepareS3StagingFile(String stagingFolder, String s3FileName) {
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3StreamCopier.java:158]: yes, it had to be done. the file name must include the schema. such a structure existed before. I moved it from the constructor to a separate method.

**Comment by andriikorotkov** [comment]
@danieldiamond, I slightly changed the formation of the name for the s3 file and added logs. you can see them in the description. If in your opinion everything looks good - I will merge this pull request.

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java [hunk 2]
```diff
@@ -44,7 +50,7 @@ public SnowflakeS3StreamCopier(String stagingFolder,
                                  S3Config s3Config,
                                  ExtendedNameTransformer nameTransformer,
                                  SqlOperations sqlOperations) {
-    super(stagingFolder, destSyncMode, schema, streamName, client, db, s3Config, nameTransformer, sqlOperations);
+    super(stagingFolder, destSyncMode, schema, streamName,  Strings.addRandomSuffix("", "", 3) + streamName, client, db, s3Config, nameTransformer, sqlOperations);
```
Comment (danieldiamond) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java:53]: Should we add a _ or other separator between the random suffix so it's easier to read

**Review by danieldiamond — APPROVED** [review]
Thanks for resolving this. One suggestion for readability

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java (hunk 2 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java]: Yes, sure. I have already added these changes

**Review by sherifnada — APPROVED** [review]
@andriikorotkov is there a test we can add to make sure there are no regressions? 

Could you also make an issue to verify this isn't happening on other Dbs? 

Both of these can be handled after releasing this PR since it's a critical fix

**Comment by sherifnada** [comment]
@andriikorotkov I've published a version to unblock @danieldiamond -- so we can probably take care of the unit test in this PR and create the follow up issue for other DBs

**Comment by andriikorotkov** [comment]
@sherifnada I have reproduced and corrected this bug for redshift as well. I also made small changes for the tests so that we don't get this error in future destinations. now the problem is in the jobs. Can I merge this pull request if not all jobs have worked correctly?

**Comment by sherifnada** [comment]
@andriikorotkov yup you can move forward

**Pull request merged**